### ISSUE TITLE
get nan Avg. Gradient when handing data with (2000,50), (2000,150)

### ISSUE THREAD

**Issue by pobingwanghai** [CannyLab/tsne-cuda#34]
efficiency test uncheck
test datapoint of (5000,512)
time:2.29
test datapoint of (5000,5)
time=2.06
test datapoint of (5000,1024)
time:2.48
test datapoint of (50,50)
time:0.66
test datapoint of (500,50)
test datapoint of (2000,50)
time: bug infinite iteration
test datapoint of (50000,50)
time:19.85s
test datapoint of (50000,256)
time:20.67s
test datapoint of (400,25)
time:0.8s

I notice that the program reports a nan Avg. Gradient when testing with data having the shape of  (2000,50), and jump to an infinite iteration when testing with data having the shape of  (500,50).

**Comment by DavidMChan** [comment]
Can you please post some code to reproduce your error? It seems very slow for the points your are computing.

**Comment by pobingwanghai** [comment]
code:
from tsnecuda import TSNE
import numpy as np
import time
X = np.random.random((5000, 300))
start=time.time()
Y=TSNE( n_components=2,verbose=1).fit_transform(X)
print(time.time()-start)
output:
first time run:3.9469144344329834
after the first time:2.3629941940307617

**Comment by pobingwanghai** [comment]
code:
from tsnecuda import TSNE
import numpy as np
import time
X = np.random.random((500, 300))
start=time.time()
Y=TSNE( n_components=2,verbose=1).fit_transform(X)
print(time.time()-start)
output: 
***,-1
infinite iteration
also try "X = np.random.random((500, 30))", but still leads to the same result.
try"X = np.random.random((3000, 50))", output 
714, -1
580, -1
2500, -1
2913, -1
[Step 0] Avg. Gradient Norm: nan
GPUassert: an illegal memory access was encountered /home/rmrao/miniconda3/conda-bld/tsnecuda_1538622901232/work/src/util/cuda_utils.cu 55

**Comment by rmrao** [comment]
This was an issue with the Barnes-Hut implementation and should be fixed by #39, which was incorporated into the [new release](https://github.com/CannyLab/tsne-cuda/releases/tag/2.0).

This hasn't been incorporated into conda yet, but we are working towards that. In the meantime, you should be able to build from source on the master branch.

**Comment by pobingwanghai** [comment]
@rmrao thank you, i got it.

**Comment by rmrao** [comment]
New conda release for CUDA 9.0 is out. You can install with `conda install tsnecuda cuda90 -c cannylab`. This should fix the issue.

**Comment by pobingwanghai** [comment]
@rmrao may i ask, whether the cuda90 version is the same as cuda100 version which requires libc 2.27

**Comment by rmrao** [comment]
Yes if you install from conda. I’m currently trying to fix this issue.

If you compile from source you should have any problems.

-Roshan

**Comment by pobingwanghai** [comment]
build from source will solve every problem.

**Comment by LucaCappelletti94** [comment]
I have just built from source and encountered this problem. How may I proceed?

**Comment by LucaCappelletti94** [comment]
Most amazingly, it only happens for values of perplexity lower or equal to 15. It runs otherwise!

**Comment by lidongyv** [comment]
@LucaCappelletti94 yep, that's true, it is still amazing that some numbers just lead to an error but some do not ...

**Comment by LucaCappelletti94** [comment]
Would it be helpful if I tried experimentally via a grid search which combinations lead to a NaN?

**Comment by LucaCappelletti94** [comment]
I have re-compiled this in a clean VM and I could not obtain any NaN after executing 3 iterations with different seed with perplexity from 1 to 100 (a total of 300 TSNEs).
